### PR TITLE
Introduce --enable_training_ops build flag

### PR DESCRIPTION
**Pull request by cianmhayes**
**Description**: Introduces a new build flag to build only the training operators without the rest of the training support.

**Motivation and Context**
Currently the --minimal_build and --enable_trianing build flags don't work well together. Much of the session and graph manipulation code in the training support has dependencies on functions that are explicitly removed from the Model, Graph, and Node classes. This change allows us to build only the training operators so they can be included in a minimal build.

### CODE REVIEW

File: orttraining/orttraining/training_ops/cpu/cpu_training_kernels.cc [hunk 1]
```diff
@@ -58,10 +58,12 @@ class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kMSDomain, 1,
 class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kMSDomain, 1, double_float, DropoutGrad);
 class ONNX_OPERATOR_TYPED_KERNEL_CLASS_NAME(kCpuExecutionProvider, kMSDomain, 1, double_double, DropoutGrad);
 
+#ifdef ENABLE_TRAINING
 class ONNX_OPERATOR_KERNEL_CLASS_NAME(kCpuExecutionProvider, kMSDomain, 1, SummaryScalar);
 class ONNX_OPERATOR_KERNEL_CLASS_NAME(kCpuExecutionProvider, kMSDomain, 1, SummaryHistogram);
 class ONNX_OPERATOR_KERNEL_CLASS_NAME(kCpuExecutionProvider, kMSDomain, 1, SummaryMerge);
 class ONNX_OPERATOR_KERNEL_CLASS_NAME(kCpuExecutionProvider, kMSDomain, 1, SummaryText);
+#endif
 class ONNX_OPERATOR_KERNEL_CLASS_NAME(kCpuExecutionProvider, kMSDomain, 1, GistBinarizeEncoder);
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cpu/cpu_training_kernels.cc:67]: Could you please include GistBinarizeEncoder/Decoder in the TRAINING-only scope? 
I don't think you will need it for your scenario.

File: onnxruntime/core/framework/utils.h [hunk 2]
```diff
@@ -156,7 +156,7 @@ constexpr ONNXTensorElementDataType GetONNXTensorElementDataType<uint64_t>() {
 
 int32_t ONNXTensorElementDataTypeToProtoTensorType(ONNXTensorElementDataType);
 
-#ifdef ENABLE_TRAINING
+#if defined(ENABLE_TRAINING) || defined(ENABLE_TRAINING_OPS)
```
Comment (SherlockNoMad) [inline onnxruntime/core/framework/utils.h:159]: Ops that need Contiguous inputs are NcclAllReduce, NcclReduceScatter, NcclAllGather... 
I don't think it need defined(ENABLE_TRAINING_OPS) here.

File: onnxruntime/core/framework/utils.cc [hunk 3]
```diff
@@ -568,7 +568,7 @@ int32_t ONNXTensorElementDataTypeToProtoTensorType(ONNXTensorElementDataType onn
   }
 }
 
-#ifdef ENABLE_TRAINING
+#if defined(ENABLE_TRAINING) || defined(ENABLE_TRAINING_OPS)
```
Comment (SherlockNoMad) [inline onnxruntime/core/framework/utils.cc:571]: Ops that need Contiguous inputs are NcclAllReduce, NcclReduceScatter, NcclAllGather... 
I don't think it need defined(ENABLE_TRAINING_OPS) here.

File: onnxruntime/core/framework/sequential_executor.cc [hunk 4]
```diff
@@ -307,7 +307,7 @@ Status SequentialExecutor::Execute(const SessionState& session_state, const std:
       node_compute_range.Begin();
 #endif
       ORT_TRY {
-#ifdef ENABLE_TRAINING
+#if defined(ENABLE_TRAINING) || defined(ENABLE_TRAINING_OPS)
```
Comment (SherlockNoMad) [inline onnxruntime/core/framework/sequential_executor.cc:310]: Ops that need Contiguous inputs are NcclAllReduce, NcclReduceScatter, NcclAllGather... 
I don't think it need defined(ENABLE_TRAINING_OPS) here.

File: onnxruntime/core/framework/parallel_executor.cc [hunk 5]
```diff
@@ -191,7 +191,7 @@ Status ParallelExecutor::RunNodeAsync(size_t p_node_index,
 
     // Execute the kernel.
     ORT_TRY {
-#ifdef ENABLE_TRAINING
+#if defined(ENABLE_TRAINING) || defined(ENABLE_TRAINING_OPS)
```
Comment (SherlockNoMad) [inline onnxruntime/core/framework/parallel_executor.cc:194]: and here

File: cmake/onnxruntime_session.cmake [hunk 6]
```diff
@@ -21,6 +21,6 @@ set_target_properties(onnxruntime_session PROPERTIES FOLDER "ONNXRuntime")
 if (onnxruntime_USE_CUDA)
   target_include_directories(onnxruntime_session PRIVATE ${onnxruntime_CUDNN_HOME}/include ${CMAKE_CUDA_TOOLKIT_INCLUDE_DIRECTORIES})
 endif()
-if (onnxruntime_ENABLE_TRAINING)
+if (onnxruntime_ENABLE_TRAINING OR onnxruntime_ENABLE_TRAINING_OPS)
   target_include_directories(onnxruntime_session PRIVATE ${ORTTRAINING_ROOT})
```
Comment (SherlockNoMad) [inline cmake/onnxruntime_session.cmake:25]: this is for TrainingSession... 
I suppose you won't need this.

File: cmake/CMakeLists.txt [hunk 7]
```diff
@@ -1242,6 +1243,10 @@ if (onnxruntime_USE_DML)
   include(dml)
 endif()
 
+if (onnxruntime_ENABLE_TRAINING_OPS)
+  add_compile_definitions(ENABLE_TRAINING_OPS)
+endif()
+
 if (onnxruntime_ENABLE_TRAINING)
   add_compile_definitions(ENABLE_TRAINING)
```
Comment (SherlockNoMad) [inline cmake/CMakeLists.txt:1252]: should we add add_compile_definitions(ENABLE_TRAINING_OPS) here to enforce 
enable_training is a strict superset of enable_training_ops.

File: cmake/onnxruntime_providers.cmake [hunk 8]
```diff
@@ -977,7 +993,7 @@ if (onnxruntime_USE_ROCM)
     list(APPEND onnxruntime_providers_rocm_src ${onnxruntime_rocm_generated_contrib_ops_cc_srcs} ${onnxruntime_rocm_generated_contrib_ops_cu_srcs})
   endif()
 
-  if (onnxruntime_ENABLE_TRAINING)
+  if (onnxruntime_ENABLE_TRAINING OR onnxruntime_ENABLE_TRAINING_OPS)
```
Comment (SherlockNoMad) [inline cmake/onnxruntime_providers.cmake:996]: Do you need rocm kernels? 
This are for AMD gpus

File: onnxruntime/core/framework/allocation_planner.cc [hunk 9]
```diff
@@ -796,7 +796,7 @@ class PlannerImpl {
     }
     return Status::OK();
   }
-#ifdef ENABLE_TRAINING
+#if defined(ENABLE_TRAINING) || defined(ENABLE_TRAINING_OPS)
```
Comment (SherlockNoMad) [inline onnxruntime/core/framework/allocation_planner.cc:799]: same here

File: cmake/CMakeLists.txt (hunk 7 above)
Comment (cianmhayes) [reply cmake/CMakeLists.txt:1256]: I added it, but now that I look at it again I may have put it in a goofy place. Let me know if you want me to move it.

File: cmake/onnxruntime_providers.cmake (hunk 8 above)
Comment (cianmhayes) [reply cmake/onnxruntime_providers.cmake]: nope, removed

File: cmake/CMakeLists.txt [hunk 10]
```diff
@@ -179,6 +180,10 @@ if (onnxruntime_ENABLE_LANGUAGE_INTEROP_OPS)
   add_compile_definitions(ENABLE_LANGUAGE_INTEROP_OPS)
 endif()
 
+if (onnxruntime_ENABLE_TRAINING)
+  add_compile_definitions(ENABLE_TRAINING_OPS)
```
Comment (SherlockNoMad) [inline cmake/CMakeLists.txt:184]: could you please move this to line 1255, there is already a if block for onnxruntime_ENABLE_TRAINING

File: cmake/onnxruntime_providers.cmake [hunk 11]
```diff
@@ -129,6 +129,19 @@ if (onnxruntime_USE_FEATURIZERS)
...
+if (onnxruntime_ENABLE_TRAINING_OPS)
+  file(GLOB_RECURSE onnxruntime_cpu_training_ops_srcs CONFIGURE_DEPENDS
+    "${ORTTRAINING_SOURCE_DIR}/training_ops/cpu/*.h"
+    "${ORTTRAINING_SOURCE_DIR}/training_ops/cpu/*.cc"
+  )
+
+  source_group(TREE ${ORTTRAINING_ROOT}/ FILES ${onnxruntime_cpu_training_ops_srcs})
+  list(APPEND onnxruntime_providers_src ${onnxruntime_cpu_training_ops_srcs})
+
+  list(REMOVE_ITEM onnxruntime_providers_src "${ORTTRAINING_SOURCE_DIR}/training_ops/cpu/tensorboard/summary_op.cc")
```
Comment (SherlockNoMad) [inline cmake/onnxruntime_providers.cmake:141]: do you need to remove gist* files here?

**Review by SherlockNoMad — APPROVED** [review]

**Review by mrry — APPROVED** [review]

**Pull request merged**